src/storage/s3.rs

```diff
@@ -0,0 +1,244 @@
+            .expect("wanted GetObject")
+        {
+            RusotoError::Unknown(http) => http.status,
+            _ => panic!("wrong error"),
+        };
+        assert_eq!(status, 404);
+    }
+
+    #[test]
+    fn test_parse_timespec() {
+        crate::test::wrapper(|_| {
+            // Test valid conversions
+            assert_eq!(
+                parse_timespec("Thu, 1 Jan 1970 00:00:00 GMT")?,
+                Timespec::new(0, 0)
+            );
+            assert_eq!(
+                parse_timespec("Mon, 16 Apr 2018 04:33:50 GMT")?,
+                Timespec::new(1523853230, 0)
+            );
+
+            // Test invalid conversion
+            assert!(parse_timespec("foo").is_err());
+
+            Ok(())
+        })
+    }
+
+    #[test]
+    fn test_path_get() {
+        crate::test::wrapper(|env| {
+            let blob = Blob {
+                path: "dir/foo.txt".into(),
+                mime: "text/plain".into(),
+                date_updated: Timespec::new(42, 0),
+                content: "Hello world!".into(),
+            };
+
+            // Add a test file to the database
+            let s3 = env.s3().upload(blob.clone());
+
+            let backend = S3Backend::new(s3.client, &s3.bucket);
+
+            // Test that the proper file was returned
+            assert_eq!(blob, backend.get("dir/foo.txt")?);
+
+            // Test that other files are not returned
+            assert_s3_404(&env, "dir/bar.txt");
+            assert_s3_404(&env, "foo.txt");
+
+            Ok(())
+        });
+    }
+}
+
+pub(crate) static S3_BUCKET_NAME: &str = "rust-docs-rs";
+
+pub(crate) fn s3_client() -> Option<S3Client> {
+    // If AWS keys aren't configured, then presume we should use the DB exclusively
+    // for file storage.
+    if std::env::var_os("AWS_ACCESS_KEY_ID").is_none() && std::env::var_os("FORCE_S3").is_none() {
+        return None;
+    }
+    let creds = match DefaultCredentialsProvider::new() {
+        Ok(creds) => creds,
+        Err(err) => {
+            warn!("failed to retrieve AWS credentials: {}", err);
+            return None;
+        }
+    };
+    Some(S3Client::new_with(
+        rusoto_core::request::HttpClient::new().unwrap(),
+        creds,
+        std::env::var("S3_ENDPOINT")
+            .ok()
+            .map(|e| Region::Custom {
+                name: std::env::var("S3_REGION").unwrap_or_else(|_| "us-west-1".to_owned()),
+                endpoint: e,
+            })
+            .unwrap_or(Region::UsWest1),
+    ))
+}
+
+pub fn move_to_s3(conn: &Connection, n: usize) -> Result<usize, Error> {
+    let trans = conn.transaction()?;
+    let client = s3_client().expect("configured s3");
+
+    let rows = trans.query(
+        &format!(
+            "SELECT path, mime, content FROM files WHERE content != E'in-s3' LIMIT {}",
+            n
+        ),
+        &[],
+    )?;
+    let count = rows.len();
+
+    let mut rt = ::tokio::runtime::Runtime::new().unwrap();
```

Comment:
In the 2018 edition the `::` syntax isn't necessary 
```suggestion
    let mut rt = tokio::runtime::Runtime::new().unwrap();
```